packages/@aws-cdk/custom-resources/lib/provider-framework/state-machine.ts

```diff
@@ -0,0 +1,80 @@
+import { Grant, IGrantable, PolicyStatement, Role, ServicePrincipal } from '@aws-cdk/aws-iam';
+import { IFunction } from '@aws-cdk/aws-lambda';
+import { CfnResource, Construct, Duration } from '@aws-cdk/core';
+
+export interface StateMachineProps {
+  readonly isCompleteHandler: IFunction;
+
+  readonly timeoutHandler: IFunction;
+
+  readonly interval: Duration;
+
+  readonly maxAttempts: number;
+
+  readonly backoffRate: number;
+}
+
+/**
+ * A very simple StateMachine construct highly customized to the provider framework.
+ * This is so that this package does not need to depend on aws-stepfunctions module.
+ */
+export class StateMachine extends Construct {
```

Comment:
To avoid confusion, I would not use a general name such as "state machine". Perhaps something like `WaiterStateMachine` or `WaiterStepFunction`.